File: DevExpress-Demo/VB/SankeyDemo.Wpf/Utils/NodeComparers.vb
Imports DevExpress.Xpf.Charts.Sankey
Imports System
Imports System.Collections.Generic

Namespace SankeyDemo

    Public MustInherit Class AscendingOrDescendingNodeComparer
        Implements IComparer(Of SankeyNode)

        Public Property Ascending As Boolean

        Public MustOverride Function Compare(ByVal x As SankeyNode, ByVal y As SankeyNode) As Integer Implements IComparer(Of SankeyNode).Compare
    End Class

    Public Class TotalWeightComparer
        Inherits AscendingOrDescendingNodeComparer

        Public Overrides Function Compare(ByVal x As SankeyNode, ByVal y As SankeyNode) As Integer
            Return If(Ascending, 1, -1) * Math.Sign(x.TotalWeight - y.TotalWeight)
        End Function

        Public Overrides Function ToString() As String
            Return "Total Weight"
        End Function
    End Class

    Public Class OutputLinkCountComparer
        Inherits AscendingOrDescendingNodeComparer
        Implements IComparer(Of SankeyNode)

        Public Overrides Function Compare(ByVal x As SankeyNode, ByVal y As SankeyNode) As Integer Implements IComparer(Of SankeyNode).Compare
            If x.OutputLinks.Count <> 0 Then
                Return If(Ascending, 1, -1) * Math.Sign(x.OutputLinks.Count - y.OutputLinks.Count)
            Else
                Return If(Ascending, 1, -1) * Math.Sign(x.InputLinks.Count - y.InputLinks.Count)
            End If
        End Function

        Public Overrides Function ToString() As String
            Return "Output Link Count"
        End Function
    End Class

    Public Class NodeNameComparer
        Inherits AscendingOrDescendingNodeComparer
        Implements IComparer(Of SankeyNode)

        Public Overrides Function Compare(ByVal x As SankeyNode, ByVal y As SankeyNode) As Integer Implements IComparer(Of SankeyNode).Compare
            Return If(Ascending, 1, -1) * x.Tag.ToString().CompareTo(y.Tag.ToString())
        End Function

        Public Overrides Function ToString() As String
            Return "Node Name"
        End Function
    End Class
End Namespace

File: DevExpress-Demo/CS/SankeyDemo.Wpf/Utils/NodeComparers.cs
using DevExpress.Xpf.Charts.Sankey;
using System;
using System.Collections.Generic;

namespace SankeyDemo {
    public abstract class AscendingOrDescendingNodeComparer : IComparer<SankeyNode> {
        public bool Ascending { get; set; }

        public abstract int Compare(SankeyNode x, SankeyNode y);
    }
    public class TotalWeightComparer : AscendingOrDescendingNodeComparer {
        public override int Compare(SankeyNode x, SankeyNode y) {
            return (Ascending ? 1 : -1) * Math.Sign(x.TotalWeight - y.TotalWeight);
        }

        public override string ToString() {
            return "Total Weight";
        }
    }

    public class OutputLinkCountComparer : AscendingOrDescendingNodeComparer, IComparer<SankeyNode> {
        public override int Compare(SankeyNode x, SankeyNode y) {
            if(x.OutputLinks.Count != 0)
                return (Ascending ? 1 : -1) * Math.Sign(x.OutputLinks.Count - y.OutputLinks.Count);
            else
                return (Ascending ? 1 : -1) * Math.Sign(x.InputLinks.Count - y.InputLinks.Count);
        }

        public override string ToString() {
            return "Output Link Count";
        }
    }

    public class NodeNameComparer : AscendingOrDescendingNodeComparer, IComparer<SankeyNode> {
        public override int Compare(SankeyNode x, SankeyNode y) {
            return (Ascending ? 1 : -1) * x.Tag.ToString().CompareTo(y.Tag.ToString());
        }

        public override string ToString() {
            return "Node Name";
        }
    }
}
